eZ/Publish/Core/Repository/BookmarkService.php

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * @var \eZ\Publish\SPI\Persistence\Bookmark\Handler
+     */
+    protected $bookmarkHandler;
+
+    /**
+     * BookmarkService constructor.
+     *
+     * @param \eZ\Publish\API\Repository\Repository $repository
+     * @param \eZ\Publish\SPI\Persistence\Bookmark\Handler $bookmarkHandler
+     */
+    public function __construct(RepositoryInterface $repository, BookmarkHandler $bookmarkHandler)
+    {
+        $this->repository = $repository;
+        $this->bookmarkHandler = $bookmarkHandler;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createBookmark(Location $location): void
+    {
+        $loadedLocation = $this->repository->getLocationService()->loadLocation($location->id);
+
+        if ($this->isBookmarked($loadedLocation)) {
+            throw new InvalidArgumentException('$location', 'location is already bookmarked.');
+        }
+
+        $createStruct = new CreateStruct();
+        $createStruct->name = $loadedLocation->contentInfo->name;
+        $createStruct->locationId = $loadedLocation->id;
+        $createStruct->userId = $this->getCurrentUserId();
+
+        $this->repository->beginTransaction();
+        try {
+            $this->bookmarkHandler->create($createStruct);
+            $this->repository->commit();
+        } catch (Exception $ex) {
+            $this->repository->rollback();
+            throw $ex;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deleteBookmark(Location $location): void
+    {
+        $loadedLocation = $this->repository->getLocationService()->loadLocation($location->id);
+
+        try {
+            $bookmark = $this->bookmarkHandler->loadByUserIdAndLocationId($this->getCurrentUserId(), $loadedLocation->id);
+        } catch (NotFoundException $e) {
+            throw new InvalidArgumentException('$location', 'location is not bookmarked.');
+        }
+
+        $this->repository->beginTransaction();
+        try {
+            $this->bookmarkHandler->delete($bookmark->id);
+            $this->repository->commit();
+        } catch (Exception $ex) {
+            $this->repository->rollback();
+            throw $ex;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadBookmarks(int $offset = 0, int $limit = 25): BookmarkList
+    {
+        $currentUserId = $this->getCurrentUserId();
+
+        $list = new BookmarkList();
+        $list->totalCount = $this->bookmarkHandler->countUserBookmarks($currentUserId);
+        if ($list->totalCount > 0) {
+            $bookmarks = $this->bookmarkHandler->loadUserBookmarks($currentUserId, $offset, $limit);
+
+            $list->items = array_map(function (Bookmark $bookmark) {
+                return $this->repository->getLocationService()->loadLocation($bookmark->locationId);
+            }, $bookmarks);
+        }
+
+        return $list;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isBookmarked(Location $location): bool
```

Comment:
> this method looks kinda fishy.
(returning values depends on exception thrown or not, seems like exception driven logic for me).

other solution would be making dedicated persistence handler method for that and handle code redundancy on a Gateway level. I'm not sure if we have time for this ;) Current approach is not so bad because it reuses existing persistence cache layer. Leaving this decision up to @adamwojs